日報の言及機能を作る

naokinaokiboo opened the pull request:
「日報の言及機能を作る」の課題内容をpushしました。
レビューよろしくお願い致します。🙇‍♂️

naokinaokiboo committed:
モデルReportMentionsを追加、マイグレーションを実行

naokinaokiboo committed:
日報の言及機能を追加

naokinaokiboo committed:
ReportMentionモデルにバリデーションを追加

naokinaokiboo committed:
日報の本文テキストのURLをリンクにする

naokinaokiboo committed:
i18nによる日本語表示

naokinaokiboo committed:
言及している日報の表示部分についてCSS対応(間隔調整のみ)

naokinaokiboo committed:
rubocopの指摘に対応

naokinaokiboo committed:
ReportMentionsテーブルにNOT NULL制約を追加

maedana commented on app/models/report_mention.rb:
```diff
@@ -0,0 +1,9 @@
+  belongs_to :mentioning_report, class_name: :Report
+  belongs_to :mentioned_report, class_name: :Report
+
+  validates :mentioning_report_id, presence: true, uniqueness: { scope: :mentioned_report_id }
```
`belongs_to`があれば自動的に必須のvalidationがかかるので`presence: true`は冗長ですね。

maedana commented on app/models/report_mention.rb:
```diff
@@ -0,0 +1,9 @@
+  belongs_to :mentioned_report, class_name: :Report
+
+  validates :mentioning_report_id, presence: true, uniqueness: { scope: :mentioned_report_id }
+  validates :mentioned_report_id, presence: true
```
同上です

maedana commented on app/views/reports/_mention.html.erb:
```diff
@@ -0,0 +1,13 @@
+      <% end %>
+    </ul>
+  <% end %>
+</div>
```
複数あるものを表示viewですし、ファイル名を複数形にしたほうがわかりやすいんじゃないかなと思いました。

naokinaokiboo replied:
確かにその通りですね😳
複数形に修正致します。

maedana commented on app/views/reports/_mention.html.erb:
```diff
@@ -0,0 +1,13 @@
+<div class="mentions-container">
+  <% unless reports.empty? %>
```
可能なかぎりifを使うほうが読みやすいかなーと思います。`present?`とか`exists?`使ってifでかいてみましょう

maedana commented on app/helpers/application_helper.rb:
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'uri'
```
これなくても動くのではないですかね

naokinaokiboo replied:
仰るように、なくても動きましたので、削除致します。😅

maedana commented on app/helpers/application_helper.rb:
```diff
@@ -18,4 +20,12 @@ def i18n_error_count(count)
+  def convert_to_linked_text(text)
+    URI.extract(text, %w[http https]).uniq.each do |url|
+      linked_url = "<a href=\"#{url}\">#{url}</a>"
+      text.gsub!(url, linked_url)
```
引数を破壊的に書き換えるのはあまりよくないこととされているのでなるべく避けたいですねー。副作用のあるメソッドなんていいますね。
```erb
    <%- formatted_content = format_content report.content %>
    <%= convert_to_linked_text(formatted_content).html_safe %>
    <%= convert_to_linked_text(formatted_content).html_safe %>
```
のようにしたときに2回目の結果が変わってしまいますが、何度実行したとしても同じ結果を返すのが望ましいかなと思います。

naokinaokiboo replied:
> 2回目の結果が変わってしまいます

自分で書いておいてなんですが、２回目の結果が変わってしまうのは、確かに知らずに使用する側からするとワナですね。💦
破壊的な変更は極力避けるべきというのが、すごく実感できました。
（特に引数を変更するのはホントダメですね。。。）
修正致します。😅

maedana commented on app/views/reports/_report.html.erb:
```diff
@@ -6,7 +6,7 @@
   <p>
     <strong><%= Report.human_attribute_name(:content) %>:</strong>
-    <%= format_content report.content %>
+    <%= convert_to_linked_text(format_content report.content).html_safe %>
```
xss怪しいかなと思いましたけど大丈夫みたいですね。ちなみに何故大丈夫かは理解してますか?  なお、`convert_to_linked_text(report.content).html_safe` とするとxss発生しますね。

naokinaokiboo replied:
> ちなみに何故大丈夫かは理解してますか?

以下のように認識しています。
- format_contentでエスケープ処理をしているため
  - ユーザーからの入力に関しては、`safe_join`（`format_content`から呼び出し）の中で`unwrapped_html_escape`でエスケープ処理がされている

maedana commented on app/controllers/reports_controller.rb:
```diff
@@ -51,4 +78,12 @@ def set_report
+    domain_path = 'http://localhost:3000/reports/'
+    ids_to_report = text.scan(/(?<=#{domain_path})[1-9][0-9]*/).uniq.map(&:to_i)
+    ids_to_report.delete(@report.id)
+    existing_report_ids = Report.ids
```
レポートの件数がめちゃくちゃ多いときのパフォーマンスが気になりますねー。普通に`ids_to_report`のレポートをActiveRecordで探索したらいいんじゃないですかね

maedana commented on app/controllers/reports_controller.rb:
```diff
@@ -20,20 +20,47 @@ def edit; end
+          ReportMention.create!(mentioning_report_id: @report.id, mentioned_report_id: id)
+        end
+      end
+    rescue ActiveRecord::RecordInvalid => e
```
いわゆる業務エラーにあたるものを例外で処理するのはあまり適切ではないかなと思いますがいかがでしょうか?

参考
https://qiita.com/jnchito/items/3ef95ea144ed15df3637

naokinaokiboo replied:
参考のページにより、業務エラー、システムエラーの基本的な制御方針がわかりました。😳

・`rescue`は`想定外`であるシステムエラー用途
・なおかつ、システムエラー発生時にどうしても行うべき処理などが無い限りは、`rescue`しない（Railsに任せる）
・業務エラーは想定できるエラーなので、`rescue`するのはおかしい（if文で制御する）

maedana commented on app/controllers/reports_controller.rb:
```diff
@@ -20,20 +20,47 @@ def edit; end
+          ReportMention.create!(mentioning_report_id: @report.id, mentioned_report_id: id_to_be_saved)
+        end
+      end
+    rescue ActiveRecord::RecordInvalid => e
```
同上です

naokinaokiboo committed:
日報の言及リンク表示用のViewファイル名を複数形に修正

・_mention.html.erb -> _mentions.html.erb
・理由：複数の言及リンクを表示するViewであるため
・ファイル名修正に伴い、Viewを呼んでいる箇所も修正

naokinaokiboo committed:
必要のないrequireを削除

naokinaokiboo committed:
条件式の書き方を修正（unless -> if）

naokinaokiboo committed:
終了タグの抜けを修正

naokinaokiboo committed:
メソッドの副作用を除去

naokinaokiboo committed:
言及している日報のidの探索方法を修正

naokinaokiboo committed:
冗長なバリデーションの記載を削除

naokinaokiboo committed:
業務エラー発生時のrescue句を削除し、メソッドの戻り値で分岐するよう修正

naokinaokiboo committed:
日報の保存、更新のロジック部分をコントローラーからモデルに移動

naokinaokiboo commented:
レビューしていただいた内容について修正致しました。🙇‍♂️
また、以下の記事も参考にし、コントローラー内のトランザクション処理をモデルに移動しました。
https://qiita.com/jnchito/items/2b57316b39d92ca0e8dc

maedana commented:
> また、以下の記事も参考にし、コントローラー内のトランザクション処理をモデルに移動しました。

これくらいはいいかなと思って流したのですが、よい対応だと思います :+1:
